### PR TITLE
Show workflow templates from custom nodes

### LINKED ISSUE #1008: [Feature Request]: add templates from custom nodes to the template gallery
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits

### What would your feature do ?

Some custom nodes also have template workflows  ( e.g. IPAdapter plus )

https://github.com/cubiq/ComfyUI_IPAdapter_plus/tree/main/examples

It would be nice if these were also displayed in the new template gallery. 

gt

### Proposed workflow

The images could be smaller ( e.g. 6 per line )  and the gallery scrollable.        

You could use a badge to indicate the custom node name.

If the workflow is just a json ( no image ) you can use use an image / file icon. 

Once the custom node creators are aware of this feature they would also probably want to use template images.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by bezo97**
Closes #1008 feature request.
Requires the following backend PR: https://github.com/comfyanonymous/ComfyUI/pull/6193 (merged)

image

Implementation details:
- Categories are provided by the api, which are the custom_nodes module names. The ones without template files are omitted.
- The first category holds the existing templates that are provided by the frontend
- Added a carousel to display more items
- Workflow thumbnails are shown when present, otherwise a generic document icon is displayed
- Using a store to avoid repeated requests. A spinner is shown while loading items the first time (but the default category items are visible anyway): 
image
- Getting the carousel to look fine on smaller screens was rather problematic but here is a best effort 
image

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-2032-Show-workflow-templates-from-custom-nodes-1666d73d365081afb9a2f2f8f53470dc) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by bezo97** [comment]
Waiting on https://github.com/comfyanonymous/ComfyUI/pull/6193

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 1]
```diff
@@ -1,68 +1,127 @@
...
+  let json
+  if (selectedTab.value.moduleName === 'default') {
+    // Default templates provided by frontend are served on this separate endpoint
+    json = await fetch(api.fileURL(`templates/${id}.json`)).then((r) =>
+      r.json()
+    )
+  } else {
+    json = await fetch(
+      api.fileURL(
+        `workflow_templates/${selectedTab.value.moduleName}/${id}.json`
```
Comment (christian-byrne) [inline src/components/templates/TemplateWorkflowsContent.vue]: In the associated backend PR:

```python
webapp.add_routes([web.static('/api/workflow_templates/' + module_name, workflows_dir)])
```

It may be better to drop the `api` prefix since these are static files. Otherwise, I believe the prefix must be added manually here in the `api.fileURL` call 👍.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 2]
```diff
@@ -1,68 +1,127 @@
...
+import Carousel from 'primevue/carousel'
+import Listbox from 'primevue/listbox'
 import Card from 'primevue/card'
 import ProgressSpinner from 'primevue/progressspinner'
-import { ref } from 'vue'
+import { onMounted, ref } from 'vue'
 import { app } from '@/scripts/app'
 import { api } from '@/scripts/api'
 import { useI18n } from 'vue-i18n'
+import toast from 'primevue/toast'
```
Comment (christian-byrne) [inline src/components/templates/TemplateWorkflowsContent.vue]: I think you need the `useToast` composable instead of the `toast` component.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 3]
```diff
@@ -78,4 +137,21 @@ const loadWorkflow = async (id: string) => {
...
+}
+
+/* Fallback graphics for workflows that don't have an image. */
+img.thumbnail::before {
+  position: absolute;
+  width: 100%;
+  height: 100%;
+  background-color: var(--comfy-menu-secondary-bg);
+  color: var(--fg-color);
+  content: '🗎';
```
Comment (christian-byrne) [inline src/components/templates/TemplateWorkflowsContent.vue]: If you want, you can use same file icon as other components:

```css
    font-family: primeicons, sans-serif;
    content: '\e958';
```

The mappings are in node_modules/primeicons/primeicons.css.

**Review by christian-byrne — COMMENTED** [review]
Tested and works for me after removing `api` prefix from route.

File: src/components/templates/TemplateWorkflowsContent.vue (hunk 1 above)
Comment (bezo97) [reply src/components/templates/TemplateWorkflowsContent.vue]: I read in a code comment that the routes without the `api/` are old but left around for a while before they're removed.  that's why I put everything under `api/` and I've changed the above code to call `apiURL`, should work now

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 4]
```diff
@@ -1,68 +1,134 @@
 <template>
-  <div
-    class="flex flex-wrap content-around justify-around gap-4 mt-4"
-    data-testid="template-workflows-content"
-  >
-    <div
-      v-for="template in templates"
-      :key="template"
-      :data-testid="`template-workflow-${template}`"
```
Comment (huchenlei) [inline src/components/templates/TemplateWorkflowsContent.vue:9]: Please keep the `data-testid` for playwright test.

File: vite.config.mts [hunk 5]
```diff
@@ -111,14 +115,18 @@ export default defineConfig({
...
-        },
+        }
       },
 
       '/ws': {
         target: DEV_SERVER_COMFYUI_URL,
         ws: true
       },
 
+      '/workflow_templates': {
```
Comment (huchenlei) [inline vite.config.mts:122]: Please revert unrelated format changes.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 6]
```diff
@@ -78,4 +144,26 @@ const loadWorkflow = async (id: string) => {
...
+
+/* Fallback graphics for workflows that don't have an image. */
+img.thumbnail::before {
+  position: absolute;
+  width: 100%;
+  height: 100%;
+  background-color: var(--comfy-menu-secondary-bg);
+  color: var(--fg-color);
+  font-family: primeicons, sans-serif;
+  content: '\e958'; /* Document icon from primevue */
```
Comment (huchenlei) [inline src/components/templates/TemplateWorkflowsContent.vue]: Let's use a less hacky way to implement fallback placeholder.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 7]
```diff
@@ -1,68 +1,134 @@
...
+    })
+  }
+})
+
 const loadWorkflow = async (id: string) => {
   loading.value = id
-  const json = await fetch(api.fileURL(`templates/${id}.json`)).then((r) =>
-    r.json()
-  )
+  let json
```
Comment (huchenlei) [inline src/components/templates/TemplateWorkflowsContent.vue:111]: Let's extract the storage of custom nodes workflow templates as a Pinia store to avoid reloading everytime the dialog is opened.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 8]
```diff
@@ -1,68 +1,134 @@
...
-              class="w-64 h-64 rounded-lg object-cover"
-            />
-            <a>
+      <template #item="slotProps">
+        <Card>
+          <template #header>
+            <div class="flex items-center justify-center">
               <div
-                class="absolute top-0 left-0 w-64 h-64 overflow-hidden opacity-0 transition duration-300 ease-in-out hover:opacity-100 bg-opacity-50 bg-black flex items-center justify-center"
+                class="relative overflow-hidden rounded-lg cursor-pointer w-64 h-64"
```
Comment (huchenlei) [inline src/components/templates/TemplateWorkflowsContent.vue]: The template workflow item here should probably be extracted as a separated Vue component.

File: src/components/templates/TemplateWorkflowsContent.vue [hunk 9]
```diff
@@ -1,68 +1,134 @@
...
-  <div
-    class="flex flex-wrap content-around justify-around gap-4 mt-4"
-    data-testid="template-workflows-content"
-  >
-    <div
-      v-for="template in templates"
-      :key="template"
-      :data-testid="`template-workflow-${template}`"
+  <div class="flex h-96">
+    <Listbox
```
Comment (huchenlei) [inline src/components/templates/TemplateWorkflowsContent.vue]: Listbox allows null value on `v-model:modeValue`. Currently deselect item on that listbox will cause error.

image

**Review by huchenlei — CHANGES_REQUESTED** [review]

**Review by huchenlei — APPROVED** [review]
